src/pointer.rs

```diff
@@ -0,0 +1,176 @@
+use crossbeam_epoch::Guard;
+use std::marker::PhantomData;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use utilities::{Cast, OverflowArithmetic};
+/// `AtomicPtr` is a pointer which can only be manipulated by
+/// atomic operations.
+#[derive(Debug)]
+pub struct AtomicPtr<T: ?Sized> {
+    /// `data` is the value of the atomic pointer.
+    data: AtomicUsize,
+    /// used for type inference.
+    _marker: PhantomData<*mut T>,
+}
+
+unsafe impl<T: ?Sized + Send + Sync> Send for AtomicPtr<T> {}
+unsafe impl<T: ?Sized + Send + Sync> Sync for AtomicPtr<T> {}
+
+impl<T> AtomicPtr<T> {
+    /// `from_usize` creates an `AtomicPtr` from `usize`.
+    const fn from_usize(data: usize) -> Self {
+        Self {
+            data: AtomicUsize::new(data),
+            _marker: PhantomData,
+        }
+    }
+
+    /// `new` creates the `value` on heap and returns its `AtomicPtr`.
+    pub fn new(value: T) -> Self {
+        let b = Box::new(value);
+        let raw_ptr = Box::into_raw(b);
+        unsafe { Self::from_usize(std::mem::transmute::<*const T, usize>(raw_ptr)) }
+    }
+
+    /// `null` returns a `AtomicPtr` with a null pointer.
+    pub const fn null() -> Self {
+        Self::from_usize(0)
+    }
+
+    /// `load` atomically loads the pointer.
+    pub fn load<'g>(&self, ord: Ordering, _: &'g Guard) -> SharedPtr<'g, T> {
+        SharedPtr::from_usize(self.data.load(ord))
+    }
+
+    /// `compare_and_set` wraps the `compare_exchange` method of `AtomicUsize`.
+    pub fn compare_and_set<'g>(
+        &self,
+        current_ptr: SharedPtr<'_, T>,
+        new_ptr: SharedPtr<'_, T>,
+        ord: Ordering,
+        _: &'g Guard,
+    ) -> Result<SharedPtr<'g, T>, (SharedPtr<'g, T>, SharedPtr<'g, T>)> {
+        let new = new_ptr.as_usize();
+        // TODO: different ordering.
+        self.data
+            .compare_exchange(current_ptr.as_usize(), new, ord, ord)
+            .map(|_| SharedPtr::from_usize(current_ptr.as_usize()))
+            .map_err(|current| (SharedPtr::from_usize(current), SharedPtr::from_usize(new)))
+    }
+}
+
+/// `SharedPtr` is a pointer which can be shared by multi-threads.
+/// `SharedPtr` can only be used with 64bit-wide pointer, and the
+/// pointer address must be 4-byte aligned.
+pub struct SharedPtr<'g, T: 'g> {
```

Comment:
should `SharedPtr` be `pub` or `pub(crate)`?

since this is a library, be careful to public structures and functions.